server/schedulers/grant_leader.go

```diff
@@ -51,46 +55,124 @@ func init() {
 			if err != nil {
 				return errors.WithStack(err)
 			}
-			conf.StoreID = id
-			conf.Name = fmt.Sprintf("%s-%d", GrantLeaderName, id)
-			conf.Ranges = ranges
+			conf.StoreIDWitRanges[id] = ranges
 			return nil
 		}
 	})
 
 	schedule.RegisterScheduler(GrantLeaderType, func(opController *schedule.OperatorController, storage *core.Storage, decoder schedule.ConfigDecoder) (schedule.Scheduler, error) {
-		conf := &grandLeaderConfig{}
+		conf := &grantLeaderSchedulerConfig{StoreIDWitRanges: make(map[uint64][]core.KeyRange), storage: storage}
+		if err := decoder(conf); err != nil {
+			return nil, err
+		}
+		conf.cluster = opController.GetCluster()
 		if err := decoder(conf); err != nil {
 			return nil, err
 		}
 		return newGrantLeaderScheduler(opController, conf), nil
 	})
 }
 
-type grandLeaderConfig struct {
-	Name    string          `json:"name"`
-	StoreID uint64          `json:"store-id"`
-	Ranges  []core.KeyRange `json:"ranges"`
+type grantLeaderSchedulerConfig struct {
+	mu               sync.RWMutex
+	storage          *core.Storage
+	StoreIDWitRanges map[uint64][]core.KeyRange `json:"store-id-ranges"`
+	cluster          opt.Cluster
+}
+
+func (conf *grantLeaderSchedulerConfig) BuildWithArgs(args []string) error {
+	if len(args) != 1 {
+		return errors.New("should specify the store-id")
+	}
+
+	id, err := strconv.ParseUint(args[0], 10, 64)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	ranges, err := getKeyRanges(args[1:])
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	conf.mu.Lock()
+	defer conf.mu.Unlock()
+	conf.StoreIDWitRanges[id] = ranges
+	return nil
+}
+
+func (conf *grantLeaderSchedulerConfig) Clone() *grantLeaderSchedulerConfig {
+	conf.mu.RLock()
+	defer conf.mu.RUnlock()
+	return &grantLeaderSchedulerConfig{
+		StoreIDWitRanges: conf.StoreIDWitRanges,
+	}
+}
+
+func (conf *grantLeaderSchedulerConfig) Persist() error {
+	name := conf.getSchedulerName()
+	conf.mu.RLock()
+	defer conf.mu.RUnlock()
+	data, err := schedule.EncodeConfig(conf)
+	if err != nil {
+		return err
+	}
+	conf.storage.SaveScheduleConfig(name, data)
+	return nil
+}
+
+func (conf *grantLeaderSchedulerConfig) getSchedulerName() string {
+	return GrantLeaderName
+}
+
+func (conf *grantLeaderSchedulerConfig) getRanges(id uint64) []string {
+	conf.mu.RLock()
+	defer conf.mu.RUnlock()
+	var res []string
+	for index := range conf.StoreIDWitRanges[id] {
+		res = append(res, (string)(conf.StoreIDWitRanges[id][index].StartKey))
+		res = append(res, (string)(conf.StoreIDWitRanges[id][index].EndKey))
+	}
```

Comment:
What about letting `ranges = conf.StoreIDWitRanges[id]` first and then iterate it to avoid the access cost of the hashmap.

Reply:
Would you like to take this idea? Your original implementation is OK for me as well.

Reply:
I have fixed it. @shafreeck 
```go
ranges := conf.StoreIDWithRanges[id]
for index := range ranges {
	res = append(res, (string)(ranges[index].StartKey))
	res = append(res, (string)(ranges[index].EndKey))
}
```

Reply:
Cool